modules/Bio/EnsEMBL/Compara/RunnableDB/ProteinTrees/HomologyGroupingFactory.pm

```diff
@@ -83,38 +84,52 @@ sub fetch_input {
     my $self = shift @_;
 
     my $mlss_id = $self->param_required('homo_mlss_id');
-
-    my $sql_1;
-    if ( $self->param('do_gene_qc') ) {
-        $sql_1 = 'SELECT homology_id FROM homology JOIN homology_member USING (homology_id) LEFT JOIN gene_member_qc USING (seq_member_id) WHERE method_link_species_set_id = ? GROUP BY homology_id HAVING COUNT(status) = 0;';
-    }
-    else {
-        $sql_1 = 'SELECT homology_id FROM homology WHERE method_link_species_set_id = ? AND description != "gene_split" ORDER BY homology_id';
+    my $mlss = $self->compara_dba->get_MethodLinkSpeciesSetAdaptor->fetch_by_dbID($mlss_id);
+    my $genome_dbs = $mlss->species_set->genome_dbs;
+
+    my @homology_ids;
+    my $homology_flatfile = $self->param_required('homology_flatfile');
+    my $do_gene_qc = $self->param('do_gene_qc');
+    my $qc_status;
+    if ( $do_gene_qc ) {
+        my $sql = "SELECT seq_member_id, status FROM gene_member_qc WHERE genome_db_id IN (?, ?)";
+        my $sth = $self->compara_dba->dbc->prepare($sql);
+        $sth->execute($genome_dbs->[0], $genome_dbs->[1]);
+        while( my ($stable_id, $status) = $sth->fetchrow() ) {
+            $qc_status->{$stable_id} = $status;
+        }
     }
-    my $sth_1 = $self->compara_dba->dbc->prepare($sql_1);
-
-    my @homology_ids = ();
-    $sth_1->execute($mlss_id);
-    while( my ($homology_id) = $sth_1->fetchrow() ) {
+    
+    open( my $hffh, '<', $homology_flatfile ) or die "Cannot open $homology_flatfile for reading";
+    my $hheader = <$hffh>;
+    my @hhead_cols = split(/\s+/, $hheader);
+    while( my $line = <$hffh> ) {
+        my $row = map_row_to_header( $line, \@hhead_cols );
+        my ( $homology_id, $homology_type, $sm_id_1, $sm_id_2 ) = ($row->{homology_id}, $row->{homology_type}, $row->{seq_member_id}, $row->{hom_seq_member_id});
```

Comment:
You seem to do this several times in this PR. Maybe in the future we could create a Utils that, given a TSV file name, returns the headers and rows in a multilevel hash with everything split nicely?

Reply:
Yes, this is repeated, but I think we will start to eat up too much memory if we parse the whole file. I like the efficiency of taking it line by line and only saving what we need.